staging_service/import_specifications/file_parser.py

```diff
@@ -0,0 +1,97 @@
+"""
+Parse import specifications files, either CSV, TSV, or Excel, and return the contents or
+error information.
+"""
+
+from dataclasses import dataclass
+from enum import Enum
+# TODO update to C impl when fixed: https://github.com/Marco-Sulla/python-frozendict/issues/26
+from frozendict.core import frozendict
+from pathlib import Path
+from typing import Tuple, Union
+
+# TODO should get mypy working at some point
+
+PRIMITIVE_TYPE = Union[str, int, float, bool, None]
+
+
+class ErrorType(Enum):
+    """
+    The type of an error encountered when trying to parse import specifications.
+    """
+    FILE_NOT_FOUND = 0
+    PARSE_FAIL = 1
+    OTHER = 2
+
+
+@dataclass(frozen=True)
+class SpecificationSource:
+    """
+    The source of an import specification.
+
+    file - the file from which the import specification was obtained.
+    tab - the spreadsheet file tab from which the import specification was obtained, if any.
+    """
+    file: Path
+    tab: str = None
+
+
+@dataclass(frozen=True)
+class Error:
+    f"""
+    An error found while attempting to parse files.
+
+    error - the type of the error.
+    message - the error message, if any
+    source - the data source associated with the error, if any
+
+    Each error type has different required arguments:
+    {ErrorType.FILE_NOT_FOUND.name}: source
+    {ErrorType.PARSE_FAIL}: source and message
+    {ErrorType.OTHER}: message. source is optional if the error applies to a source file.
+
+    """
+    error: ErrorType
+    message: str = None
+    source: SpecificationSource = None
+
+    def __post_init__(self):
+        if self.error == ErrorType.FILE_NOT_FOUND:
+            if not self.source:
+                raise ValueError(f'source is required for a {ErrorType.FILE_NOT_FOUND.name} error')
+        elif self.error == ErrorType.PARSE_FAIL:
+            if not self.source or not self.message:
+                pf = ErrorType.PARSE_FAIL.name
+                raise ValueError(f'source and message are required for a {pf} error')
+        elif self.error == ErrorType.OTHER:
+            if not self.message:
+                raise ValueError(f'message is required for a {ErrorType.OTHER.name} error')
+        else:
+            assert 0, "unexpected error type"  # can't test this line
```

Comment:
That's true for now, but if we start using mypy I think that'll cause an error, since two `ErrorType` classes defined in different places are not the same class, IIUC